src/Features/LanguageServer/Protocol/Handler/Diagnostics/Public/PublicDocumentPullDiagnosticsHandler_IOnInitialized.cs

```diff
@@ -17,15 +18,29 @@ internal sealed partial class PublicDocumentPullDiagnosticsHandler : IOnInitiali
 {
     public async Task OnInitializedAsync(ClientCapabilities clientCapabilities, RequestContext context, CancellationToken cancellationToken)
     {
-        // Dynamically register for all of our document diagnostic sources.
+        // Dynamically register for all relevant diagnostic sources.
         if (clientCapabilities?.TextDocument?.Diagnostic?.DynamicRegistration is true)
         {
             // TODO: Hookup an option changed handler for changes to BackgroundAnalysisScopeOption
             //       to dynamically register/unregister the non-local document diagnostic source.
 
-            // Task diagnostics shouldn't be reported through VSCode (it has its own task stuff). Additional cleanup needed.
-            var sources = DiagnosticSourceManager.GetDocumentSourceProviderNames().Where(source => source != PullDiagnosticCategories.Task);
-            var registrations = sources.Select(FromSourceName).ToArray();
+            var documentSources = DiagnosticSourceManager.GetDocumentSourceProviderNames(clientCapabilities);
+            var workspaceSources = DiagnosticSourceManager.GetWorkspaceSourceProviderNames(clientCapabilities);
+
+            // All diagnostic sources have to be registered under the document pull method name,
+            // See https://github.com/microsoft/language-server-protocol/issues/1723
+            //
+            // Additionally if a source name is used by both document and workspace pull (e.g. enc)
+            // we don't want to send two registrations, instead we should send a single registration
+            // that also sets the workspace pull option.
+            //
+            // So we build up a unique set of source names and mark if each one is also a workspace source.
+            var allSources = documentSources
```

Comment:
previously the doc handler registerd doc sources and workspace handler registered wortkspace sources.  However this meant that registrations would get overwritten for sources using the same name in both doc and workspace (e.g. enc) - because vscode requires us to register them all under the same method name (with different options for workspace)